src/compiler/checker.ts

```diff
@@ -10506,6 +10887,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return localName;
             }
         }
+
+        function isUnfolding(context: NodeBuilderContext) {
+            return context.unfoldDepth !== -1;
+        }
+
+        function isHashPrivate(s: Symbol): boolean {
+            // `valueDeclaration` could be undefined if inherited from
+            // a union/intersection base type, but inherited properties
+            // don't matter here.
+            return !!s.valueDeclaration && isNamedDeclaration(s.valueDeclaration) && isPrivateIdentifier(s.valueDeclaration.name);
+        }
+
+        function getHashPrivateName(s: Symbol): PrivateIdentifier | undefined {
+            if (s.valueDeclaration && isNamedDeclaration(s.valueDeclaration) && isPrivateIdentifier(s.valueDeclaration.name)) {
+                return factory.cloneNode(s.valueDeclaration.name);
+            }
+            return undefined;
+        }
+    }
+
+    /** Returns true if a type is declared in a lib file. */
+    // Don't unfold types like `Array` or `Promise`, instead treating them as transparent.
+    function isLibType(type: Type): boolean {
+        const symbol = (getObjectFlags(type) & ObjectFlags.Reference) !== 0 ? (type as TypeReference).target.symbol : type.symbol;
+        return isTupleType(type) || !!(symbol?.declarations?.some(decl => host.isSourceFileDefaultLibrary(getSourceFileOfNode(decl))));

```

Comment:
Hm. Applying this to all lib types feels like a broad brush, but I agree that I definitely don't want to see `Array` or `Promise` expand. Meanwhile something like `AddressErrors` might still be useful to me. Is it worth whitelisting lib types in some way or making it configurable? Eh, future improvement. This is probably fine for now.